{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import allel\n",
    "import zarr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import h5py\n",
    "import sys\n",
    "import dask\n",
    "nnz = np.count_nonzero\n",
    "import malariagen_data\n",
    "ag3 = malariagen_data.Ag3(\"gs://vo_agam_release/\")\n",
    "import h5py\n",
    "from allel import chunked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log(*msg):\n",
    "    print(' '.join(map(str, msg)), file=sys.stdout)\n",
    "    sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>contributor</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>sex_call</th>\n",
       "      <th>...</th>\n",
       "      <th>pca_pops</th>\n",
       "      <th>pca_country</th>\n",
       "      <th>aim_species_fraction_colu_no2l</th>\n",
       "      <th>taxon</th>\n",
       "      <th>cohort_admin1_year</th>\n",
       "      <th>cohort_admin1_month</th>\n",
       "      <th>cohort_admin1_quarter</th>\n",
       "      <th>cohort_admin2_year</th>\n",
       "      <th>cohort_admin2_month</th>\n",
       "      <th>cohort_admin2_quarter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB0085-Cx</td>\n",
       "      <td>BF2-4</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>-4.235</td>\n",
       "      <td>11.150</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>An.gam</td>\n",
       "      <td>BFgam</td>\n",
       "      <td>0.016488</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>BF-09_gamb_2012</td>\n",
       "      <td>BF-09_gamb_2012_07</td>\n",
       "      <td>BF-09_gamb_2012_Q3</td>\n",
       "      <td>BF-09_Houet_gamb_2012</td>\n",
       "      <td>BF-09_Houet_gamb_2012_07</td>\n",
       "      <td>BF-09_Houet_gamb_2012_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB0086-Cx</td>\n",
       "      <td>BF2-6</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>-4.235</td>\n",
       "      <td>11.150</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>An.gam</td>\n",
       "      <td>BFgam</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>BF-09_gamb_2012</td>\n",
       "      <td>BF-09_gamb_2012_07</td>\n",
       "      <td>BF-09_gamb_2012_Q3</td>\n",
       "      <td>BF-09_Houet_gamb_2012</td>\n",
       "      <td>BF-09_Houet_gamb_2012_07</td>\n",
       "      <td>BF-09_Houet_gamb_2012_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AB0087-C</td>\n",
       "      <td>BF3-3</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>11.233</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>An.col</td>\n",
       "      <td>BFcol</td>\n",
       "      <td>0.981007</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>BF-09_colu_2012</td>\n",
       "      <td>BF-09_colu_2012_07</td>\n",
       "      <td>BF-09_colu_2012_Q3</td>\n",
       "      <td>BF-09_Houet_colu_2012</td>\n",
       "      <td>BF-09_Houet_colu_2012_07</td>\n",
       "      <td>BF-09_Houet_colu_2012_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB0088-C</td>\n",
       "      <td>BF3-5</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>11.233</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>An.col</td>\n",
       "      <td>BFcol</td>\n",
       "      <td>0.987593</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>BF-09_colu_2012</td>\n",
       "      <td>BF-09_colu_2012_07</td>\n",
       "      <td>BF-09_colu_2012_Q3</td>\n",
       "      <td>BF-09_Houet_colu_2012</td>\n",
       "      <td>BF-09_Houet_colu_2012_07</td>\n",
       "      <td>BF-09_Houet_colu_2012_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB0089-Cx</td>\n",
       "      <td>BF3-8</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>11.233</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>An.col</td>\n",
       "      <td>BFcol</td>\n",
       "      <td>0.973597</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>BF-09_colu_2012</td>\n",
       "      <td>BF-09_colu_2012_07</td>\n",
       "      <td>BF-09_colu_2012_Q3</td>\n",
       "      <td>BF-09_Houet_colu_2012</td>\n",
       "      <td>BF-09_Houet_colu_2012_07</td>\n",
       "      <td>BF-09_Houet_colu_2012_Q3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id partner_sample_id  contributor       country location  year   \n",
       "0  AB0085-Cx             BF2-4  Austin Burt  Burkina Faso     Pala  2012  \\\n",
       "1  AB0086-Cx             BF2-6  Austin Burt  Burkina Faso     Pala  2012   \n",
       "2   AB0087-C             BF3-3  Austin Burt  Burkina Faso     Bana  2012   \n",
       "3   AB0088-C             BF3-5  Austin Burt  Burkina Faso     Bana  2012   \n",
       "4  AB0089-Cx             BF3-8  Austin Burt  Burkina Faso     Bana  2012   \n",
       "\n",
       "   month  longitude  latitude sex_call  ... pca_pops pca_country   \n",
       "0      7     -4.235    11.150        F  ...   An.gam       BFgam  \\\n",
       "1      7     -4.235    11.150        F  ...   An.gam       BFgam   \n",
       "2      7     -4.472    11.233        F  ...   An.col       BFcol   \n",
       "3      7     -4.472    11.233        F  ...   An.col       BFcol   \n",
       "4      7     -4.472    11.233        F  ...   An.col       BFcol   \n",
       "\n",
       "   aim_species_fraction_colu_no2l     taxon cohort_admin1_year   \n",
       "0                        0.016488   gambiae    BF-09_gamb_2012  \\\n",
       "1                        0.026316   gambiae    BF-09_gamb_2012   \n",
       "2                        0.981007  coluzzii    BF-09_colu_2012   \n",
       "3                        0.987593  coluzzii    BF-09_colu_2012   \n",
       "4                        0.973597  coluzzii    BF-09_colu_2012   \n",
       "\n",
       "  cohort_admin1_month cohort_admin1_quarter     cohort_admin2_year   \n",
       "0  BF-09_gamb_2012_07    BF-09_gamb_2012_Q3  BF-09_Houet_gamb_2012  \\\n",
       "1  BF-09_gamb_2012_07    BF-09_gamb_2012_Q3  BF-09_Houet_gamb_2012   \n",
       "2  BF-09_colu_2012_07    BF-09_colu_2012_Q3  BF-09_Houet_colu_2012   \n",
       "3  BF-09_colu_2012_07    BF-09_colu_2012_Q3  BF-09_Houet_colu_2012   \n",
       "4  BF-09_colu_2012_07    BF-09_colu_2012_Q3  BF-09_Houet_colu_2012   \n",
       "\n",
       "        cohort_admin2_month     cohort_admin2_quarter  \n",
       "0  BF-09_Houet_gamb_2012_07  BF-09_Houet_gamb_2012_Q3  \n",
       "1  BF-09_Houet_gamb_2012_07  BF-09_Houet_gamb_2012_Q3  \n",
       "2  BF-09_Houet_colu_2012_07  BF-09_Houet_colu_2012_Q3  \n",
       "3  BF-09_Houet_colu_2012_07  BF-09_Houet_colu_2012_Q3  \n",
       "4  BF-09_Houet_colu_2012_07  BF-09_Houet_colu_2012_Q3  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_excel('../Data/Meta/meta_new_3_0.xlsx', sheet_name=\"meta3.0\")\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['An.gam', 'An.col', 'gcx1-GM', 'gcx2', 'gcx1-GW'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.pca_pops.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset_fn = '../Data/Final/Data/chrom_3_west_final.h5'\n",
    "callset = h5py.File(callset_fn, mode='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeDaskArray shape=(160926, 1190, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">1185</th><th style=\"text-align: center\">1186</th><th style=\"text-align: center\">1187</th><th style=\"text-align: center\">1188</th><th style=\"text-align: center\">1189</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">160923</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">160924</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">160925</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeDaskArray shape=(160926, 1190, 2) dtype=int8>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt = allel.GenotypeDaskArray(callset['3']['calldata'])\n",
    "gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset_fn = '/media/randomx/Volume/Ag1000G/Phase3/Data/Extras/outgroup_allele_counts_phase3.h5'\n",
    "callset_out = h5py.File(callset_fn, mode='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<KeysViewHDF5 ['chri', 'epir']>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset_out['3R'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_3R = pd.read_csv('../Data/Final/Data/pos_3R_prun.csv', header=None).iloc[0].to_numpy() \n",
    "pos_3L = pd.read_csv('../Data/Final/Data/pos_3L_prun.csv', header=None).iloc[0].to_numpy() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_chri_3R = callset_out['3R']['chri'][pos_3R]\n",
    "ac_chri_3L = callset_out['3L']['chri'][pos_3L]\n",
    "ac_chri = np.vstack((ac_chri_3R,ac_chri_3L))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "subpops = {\n",
    "    #'all': list(range(len(meta))),\n",
    "    'An.gam': meta[meta.pca_pops == 'An.gam'].index.tolist(),\n",
    "    'An.col': meta[meta.pca_pops == 'An.col'].index.tolist(),\n",
    "    'gcx2': meta[meta.pca_pops == 'gcx2'].index.tolist(),\n",
    "    'gcx1-GM': meta[meta.pca_pops == 'gcx1-GM'].index.tolist(),\n",
    "    'gcx1-GW': meta[meta.pca_pops == 'gcx1-GW'].index.tolist(),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21.2 ms, sys: 4.01 ms, total: 25.2 ms\n",
      "Wall time: 24.9 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ac_subpops = gt.count_alleles_subpops(subpops, max_allele=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'An.gam': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>,\n",
       " 'An.col': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>,\n",
       " 'gcx2': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>,\n",
       " 'gcx1-GM': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>,\n",
       " 'gcx1-GW': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>,\n",
       " 'chri': <AlleleCountsDaskArray shape=(160926, 2) dtype=int32>}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac_subpops['chri'] = allel.AlleleCountsDaskArray(ac_chri[:,:2])\n",
    "ac_subpops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "populations = ['chri','An.gam','An.col','gcx1-GM','gcx1-GW','gcx2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "acsu = dict()\n",
    "for pop in populations:\n",
    "    acsu[pop] = ac_subpops[pop][:, :2].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_treemix(acs, fn):\n",
    "    pops = sorted(acs.keys())\n",
    "    n_variants = acs[pops[0]].shape[0]\n",
    "    n_alleles = acs[pops[0]].shape[1]\n",
    "    assert n_alleles == 2, 'only biallelic variants supported'\n",
    "    for pop in pops[1:]:\n",
    "        assert n_variants == acs[pop].shape[0], 'bad number of variants for pop %s' % pop\n",
    "        assert n_alleles == acs[pop].shape[1], 'bad number of alleles for pop %s' % pop\n",
    "        \n",
    "    with open(fn, 'wt', encoding='ascii') as f:\n",
    "        print(' '.join(pops), file=f)\n",
    "        for i in range(n_variants):\n",
    "            print(' '.join([','.join(map(str, acs[pop][i])) for pop in pops]), file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: directory '0531_new_treemix/' creata\n",
      "0531_new_treemix/treemix_fw_chri:\t 69.6% -- replaced with 0531_new_treemix/treemix_fw_chri.gz\n"
     ]
    }
   ],
   "source": [
    "outdir = 'Treemix/'\n",
    "!mkdir -pv {outdir}\n",
    "fn = os.path.join(outdir, 'treemix_fw_chri')\n",
    "to_treemix(acsu, fn)\n",
    "!gzip -fv {fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>An.col</th>\n",
       "      <th>An.gam</th>\n",
       "      <th>chri</th>\n",
       "      <th>gcx1-GM</th>\n",
       "      <th>gcx1-GW</th>\n",
       "      <th>gcx2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>748,0</td>\n",
       "      <td>856,42</td>\n",
       "      <td>0,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>394,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>361,387</td>\n",
       "      <td>234,664</td>\n",
       "      <td>0,0</td>\n",
       "      <td>66,88</td>\n",
       "      <td>78,108</td>\n",
       "      <td>192,202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>748,0</td>\n",
       "      <td>890,6</td>\n",
       "      <td>0,0</td>\n",
       "      <td>150,4</td>\n",
       "      <td>169,17</td>\n",
       "      <td>391,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>666,80</td>\n",
       "      <td>894,1</td>\n",
       "      <td>0,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>184,2</td>\n",
       "      <td>370,22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>730,16</td>\n",
       "      <td>792,104</td>\n",
       "      <td>0,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>185,1</td>\n",
       "      <td>391,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160921</th>\n",
       "      <td>601,143</td>\n",
       "      <td>745,153</td>\n",
       "      <td>1,0</td>\n",
       "      <td>110,44</td>\n",
       "      <td>149,36</td>\n",
       "      <td>320,74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160922</th>\n",
       "      <td>675,68</td>\n",
       "      <td>778,114</td>\n",
       "      <td>1,0</td>\n",
       "      <td>145,8</td>\n",
       "      <td>166,18</td>\n",
       "      <td>362,29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160923</th>\n",
       "      <td>646,92</td>\n",
       "      <td>814,82</td>\n",
       "      <td>1,0</td>\n",
       "      <td>136,18</td>\n",
       "      <td>164,22</td>\n",
       "      <td>331,61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160924</th>\n",
       "      <td>734,14</td>\n",
       "      <td>893,3</td>\n",
       "      <td>0,0</td>\n",
       "      <td>147,7</td>\n",
       "      <td>186,0</td>\n",
       "      <td>393,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160925</th>\n",
       "      <td>732,16</td>\n",
       "      <td>895,2</td>\n",
       "      <td>1,0</td>\n",
       "      <td>149,5</td>\n",
       "      <td>184,2</td>\n",
       "      <td>385,9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160926 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         An.col   An.gam chri gcx1-GM gcx1-GW     gcx2\n",
       "0         748,0   856,42  0,0   154,0   186,0    394,0\n",
       "1       361,387  234,664  0,0   66,88  78,108  192,202\n",
       "2         748,0    890,6  0,0   150,4  169,17    391,1\n",
       "3        666,80    894,1  0,0   154,0   184,2   370,22\n",
       "4        730,16  792,104  0,0   154,0   185,1    391,1\n",
       "...         ...      ...  ...     ...     ...      ...\n",
       "160921  601,143  745,153  1,0  110,44  149,36   320,74\n",
       "160922   675,68  778,114  1,0   145,8  166,18   362,29\n",
       "160923   646,92   814,82  1,0  136,18  164,22   331,61\n",
       "160924   734,14    893,3  0,0   147,7   186,0    393,1\n",
       "160925   732,16    895,2  1,0   149,5   184,2    385,9\n",
       "\n",
       "[160926 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv('Treemix/treemix_fw_chri.gz', sep=' ')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sandbox Treemix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset_pass= ag3 = malariagen_data.Ag3(\"gs://vo_agam_release/\")\n",
    "allele_counts= zarr.open('../Data/Extras/fw_biallelic_count_country.zarr/')\n",
    "outgroup_allele_counts= zarr.open('../Data/Extras/outgroup_alleles_phase3.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sel1 = np.array([True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True])\n",
    "sel2 = np.array([True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,False,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,False,False,True,True,True,True,True,True,False,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,False,True,True,False,True,True,False,False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outgroup_ascertainment(chrom, start, stop, outgroups):\n",
    "    \n",
    "    # locate region\n",
    "    filter_pass = ag3.site_filters(chrom, mask=\"gamb_colu\")\n",
    "    f = filter_pass[:].compute()\n",
    "    pos = ag3.snp_sites(chrom, field='POS')\n",
    "    pos = allel.SortedIndex(pos[f].compute())\n",
    "    locr = pos.locate_range(start, stop)\n",
    "    \n",
    "    # ascertain SNPs\n",
    "    loca = np.zeros(pos.shape, dtype='b1')\n",
    "    loca[locr] = True\n",
    "    log('outgroup ascertainment, initial', nnz(loca))\n",
    "    for s in outgroups:\n",
    "        ac = allel.AlleleCountsArray(outgroup_allele_counts[chrom][s][:,:2])\n",
    "        # non-missing\n",
    "        locs = (ac.sum(axis=1) > 0)\n",
    "        loca &= locs\n",
    "        log(s, nnz(loca))\n",
    "        \n",
    "    return loca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ingroup_ascertainment(chrom, start, stop, segpops):\n",
    "\n",
    "    # locate region\n",
    "    filter_pass = ag3.site_filters(chrom, mask=\"gamb_colu\")\n",
    "    f = filter_pass[:].compute()\n",
    "    pos = ag3.snp_sites(chrom, field='POS')\n",
    "    pos = allel.SortedIndex(pos[f].compute())\n",
    "    locr = pos.locate_range(start, stop)\n",
    "\n",
    "    # ascertain SNPs\n",
    "    loca = np.zeros(pos.shape, dtype='b1')\n",
    "    loca[locr] = True\n",
    "    log('ingroup ascertainment, initial', nnz(loca))\n",
    "\n",
    "    \n",
    "    # require segregating\n",
    "    for pop in segpops:\n",
    "        ac = allel.AlleleCountsArray(allele_counts[chrom][pop][:])\n",
    "        loc_seg = ac.min(axis=1) > 0\n",
    "        loca &= loc_seg\n",
    "        log('after require segregating in', pop, nnz(loca))\n",
    "        \n",
    "    return loca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_treemix(acs, fn):\n",
    "    pops = sorted(acs.keys())\n",
    "    n_variants = acs[pops[0]].shape[0]\n",
    "    n_alleles = acs[pops[0]].shape[1]\n",
    "    assert n_alleles == 2, 'only biallelic variants supported'\n",
    "    for pop in pops[1:]:\n",
    "        assert n_variants == acs[pop].shape[0], 'bad number of variants for pop %s' % pop\n",
    "        assert n_alleles == acs[pop].shape[1], 'bad number of alleles for pop %s' % pop\n",
    "        \n",
    "    with open(fn, 'wt', encoding='ascii') as f:\n",
    "        print(' '.join(pops), file=f)\n",
    "        for i in range(n_variants):\n",
    "            print(' '.join([','.join(map(str, acs[pop][i])) for pop in pops]), file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def downsample_and_prune(chrom, start, stop, loc_asc,\n",
    "                         n=100000, ldp_size=500, ldp_step=250, ldp_threshold=.1, ldp_n_iter=1):\n",
    "\n",
    "    # all variant positions    \n",
    "    filter_pass = ag3.site_filters(chrom, mask=\"gamb_colu\")\n",
    "    f = filter_pass[:].compute()\n",
    "    pos = ag3.snp_sites(chrom, field='POS')\n",
    "    pos = pos[f]\n",
    "    pos = allel.SortedIndex(pos)\n",
    "    posa = pos[loc_asc]\n",
    "\n",
    "    # randomly downsample\n",
    "    gt = ag3.snp_genotypes(chrom, sample_sets=[\"AG1000G-BF-A\", \"AG1000G-BF-B\", \"AG1000G-BF-C\", \"AG1000G-CI\", \"AG1000G-GH\", \"AG1000G-GM-A\", \"AG1000G-GM-B\", \n",
    "                                       \"AG1000G-GM-C\", \"AG1000G-GN-A\", \"AG1000G-GN-B\", \"AG1000G-ML-A\", \"AG1000G-ML-B\", \"AG1000G-GW\"])\n",
    "    genotype = allel.GenotypeDaskArray(gt).subset(sel0=f) \n",
    "    genotype = genotype.subset(sel1 = sel1)\n",
    "    genotype = genotype.subset(sel1 = sel2)\n",
    "    geno_subset = genotype.subset(sel0=loc_asc).compute()\n",
    "    downsample_n = 100000\n",
    "    \n",
    "    if posa.size > downsample_n:\n",
    "        print(\"Downsampling... to {0} from {1}\".format(downsample_n, posa.size))\n",
    "        idx = np.random.choice(np.arange(0, posa.size), downsample_n, False)\n",
    "        idx.sort()\n",
    "        pos = np.take(posa, idx)\n",
    "        g = geno_subset.take(idx, axis=0)\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        # skip downsampling\n",
    "        g = geno_subset\n",
    "\n",
    " \n",
    "    gn = g.to_n_alt()\n",
    "\n",
    "    \n",
    "    # prune    \n",
    "    for i in range(ldp_n_iter):\n",
    "        loc_unlinked = allel.locate_unlinked(gn, size=ldp_size, step=ldp_step, threshold=ldp_threshold)\n",
    "        n = np.count_nonzero(loc_unlinked)\n",
    "        n_remove = gn.shape[0] - n\n",
    "        log('iteration', i+1, 'retaining', n, 'removing', n_remove, 'variants')\n",
    "        gnu = gn.compress(loc_unlinked, axis=0)\n",
    "        posu = pos.compress(loc_unlinked)\n",
    "        locu = pos.locate_keys(posu)\n",
    "\n",
    "    return locu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_analysis(rname, chrom, start, stop, outgroups, segpops,\n",
    "                 n=100000, ldp_size=500, ldp_step=250, ldp_threshold=.1, ldp_n_iter=1):\n",
    "\n",
    "    # initial ascertainment\n",
    "    loc_og_asc = outgroup_ascertainment(chrom, start, stop, outgroups=outgroups)\n",
    "    loc_ig_asc = ingroup_ascertainment(chrom, start, stop, segpops=segpops)\n",
    "    loc_asc = loc_og_asc & loc_ig_asc\n",
    "    log('initial ascertainment', nnz(loc_asc))\n",
    "    \n",
    "    # downsample and prune\n",
    "    locu = downsample_and_prune(chrom, start, stop, loc_asc, \n",
    "                                n=n, ldp_size=ldp_size, ldp_step=ldp_step, \n",
    "                                ldp_threshold=ldp_threshold, ldp_n_iter=ldp_n_iter)\n",
    "    \n",
    "    # write allele counts\n",
    "    acsu = dict()\n",
    "    for pop in segpops:\n",
    "        acsu[pop] = allele_counts[chrom][pop][:, :2][locu]\n",
    "    for pop in outgroups:\n",
    "        acsu[pop] = outgroup_allele_counts[chrom][pop][:, :2][locu]\n",
    "\n",
    "    outdir = 'treemix/'\n",
    "    !mkdir -pv {outdir}\n",
    "    fn = os.path.join(outdir, '%s.new_treemix.txt' % rname)\n",
    "    to_treemix(acsu, fn)\n",
    "    !gzip -fv {fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "outgroups = ['chri']\n",
    "segpops = ['BFgam', 'BFcol', 'CIcol', 'GHcol', 'GHgam', 'GM-2', 'GM-1','GNgam', 'GNcol', 'MAgam', 'MAcol', 'GW', 'GWgam']\n",
    "n = 100000\n",
    "ldp_n_iter = 5\n",
    "region_3R_24mbp = '3R-24Mbp', '3R', 1, 24_000_000\n",
    "region_3L_free = '3L-free', '3L', 18_000_000, 41_000_000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3R-24Mbp 3R 1 24000000\n",
      "outgroup ascertainment, initial 18316117\n",
      "chri 12088099\n",
      "ingroup ascertainment, initial 18316117\n",
      "after require segregating in BFgam 2506748\n",
      "after require segregating in BFcol 1102387\n",
      "after require segregating in CIcol 738367\n",
      "after require segregating in GHcol 616605\n",
      "after require segregating in GHgam 455698\n",
      "after require segregating in GM-2 405818\n",
      "after require segregating in GM-1 404126\n",
      "after require segregating in GNgam 396964\n",
      "after require segregating in GNcol 227928\n",
      "after require segregating in MAgam 226308\n",
      "after require segregating in MAcol 225389\n",
      "after require segregating in GW 224217\n",
      "after require segregating in GWgam 168342\n",
      "initial ascertainment 93620\n",
      "iteration 1 retaining 73990 removing 19630 variants\n",
      "iteration 2 retaining 73990 removing 19630 variants\n",
      "iteration 3 retaining 73990 removing 19630 variants\n",
      "iteration 4 retaining 73990 removing 19630 variants\n",
      "iteration 5 retaining 73990 removing 19630 variants\n",
      "treemix/3R-24Mbp.new_treemix.txt:\t 96.1% -- replaced with treemix/3R-24Mbp.new_treemix.txt.gz\n"
     ]
    }
   ],
   "source": [
    "rname, chrom, start, stop = region_3R_24mbp\n",
    "log(rname, chrom, start, stop)\n",
    "run_analysis(rname, chrom, start, stop, outgroups, segpops, n=n, ldp_n_iter=ldp_n_iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3L-free 3L 18000000 41000000\n",
      "outgroup ascertainment, initial 16952385\n",
      "chri 10040109\n",
      "ingroup ascertainment, initial 16952385\n",
      "after require segregating in BFgam 2325501\n",
      "after require segregating in BFcol 1002959\n",
      "after require segregating in CIcol 667921\n",
      "after require segregating in GHcol 559952\n",
      "after require segregating in GHgam 410666\n",
      "after require segregating in GM-2 366701\n",
      "after require segregating in GM-1 365169\n",
      "after require segregating in GNgam 358384\n",
      "after require segregating in GNcol 208480\n",
      "after require segregating in MAgam 206717\n",
      "after require segregating in MAcol 205884\n",
      "after require segregating in GW 204976\n",
      "after require segregating in GWgam 156252\n",
      "initial ascertainment 76858\n",
      "iteration 1 retaining 60885 removing 15973 variants\n",
      "iteration 2 retaining 60885 removing 15973 variants\n",
      "iteration 3 retaining 60885 removing 15973 variants\n",
      "iteration 4 retaining 60885 removing 15973 variants\n",
      "iteration 5 retaining 60885 removing 15973 variants\n",
      "treemix/3L-free.new_treemix.txt:\t 97.0% -- replaced with treemix/3L-free.new_treemix.txt.gz\n"
     ]
    }
   ],
   "source": [
    "rname, chrom, start, stop = region_3L_free\n",
    "log(rname, chrom, start, stop)\n",
    "run_analysis(rname, chrom, start, stop, outgroups, segpops, n=n, ldp_n_iter=ldp_n_iter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BFcol</th>\n",
       "      <th>BFgam</th>\n",
       "      <th>CIcol</th>\n",
       "      <th>GHcol</th>\n",
       "      <th>GHgam</th>\n",
       "      <th>GM-1</th>\n",
       "      <th>GM-2</th>\n",
       "      <th>GNcol</th>\n",
       "      <th>GNgam</th>\n",
       "      <th>GW</th>\n",
       "      <th>GWgam</th>\n",
       "      <th>MAcol</th>\n",
       "      <th>MAgam</th>\n",
       "      <th>chri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>270,0</td>\n",
       "      <td>312,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>393,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,2</td>\n",
       "      <td>150,4</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>181,5</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73985</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73986</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73987</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73988</th>\n",
       "      <td>270,0</td>\n",
       "      <td>313,1</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73989</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73990 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BFcol  BFgam  CIcol  GHcol GHgam   GM-1   GM-2 GNcol  GNgam     GW  \\\n",
       "0      270,0  312,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "1      270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "2      268,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "3      268,0  314,0  160,0  126,0  72,0  393,0  154,0  22,0  246,0  186,0   \n",
       "4      268,0  314,0  160,0  126,0  72,0  392,2  150,4  22,0  246,0  181,5   \n",
       "...      ...    ...    ...    ...   ...    ...    ...   ...    ...    ...   \n",
       "73985  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "73986  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "73987  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "73988  270,0  313,1  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "73989  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "\n",
       "      GWgam  MAcol  MAgam chri  \n",
       "0      16,0  170,0  250,0  0,0  \n",
       "1      16,0  170,0  250,0  0,0  \n",
       "2      16,0  170,0  250,0  0,0  \n",
       "3      16,0  170,0  250,0  0,0  \n",
       "4      16,0  170,0  250,0  0,0  \n",
       "...     ...    ...    ...  ...  \n",
       "73985  16,0  170,0  250,0  0,0  \n",
       "73986  16,0  170,0  250,0  0,0  \n",
       "73987  16,0  170,0  250,0  0,0  \n",
       "73988  16,0  170,0  250,0  0,0  \n",
       "73989  16,0  170,0  250,0  0,0  \n",
       "\n",
       "[73990 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1=pd.read_csv('treemix/3R-24Mbp.new_treemix.txt.gz', sep=' ')\n",
    "df_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BFcol</th>\n",
       "      <th>BFgam</th>\n",
       "      <th>CIcol</th>\n",
       "      <th>GHcol</th>\n",
       "      <th>GHgam</th>\n",
       "      <th>GM-1</th>\n",
       "      <th>GM-2</th>\n",
       "      <th>GNcol</th>\n",
       "      <th>GNgam</th>\n",
       "      <th>GW</th>\n",
       "      <th>GWgam</th>\n",
       "      <th>MAcol</th>\n",
       "      <th>MAgam</th>\n",
       "      <th>chri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>269,1</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>124,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,0</td>\n",
       "      <td>152,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>168,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>124,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,0</td>\n",
       "      <td>152,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>168,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>123,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,0</td>\n",
       "      <td>152,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>168,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>124,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,0</td>\n",
       "      <td>152,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>168,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>124,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>391,0</td>\n",
       "      <td>152,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>168,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60880</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60881</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60882</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>386,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60883</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60884</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60885 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BFcol  BFgam  CIcol  GHcol GHgam   GM-1   GM-2 GNcol  GNgam     GW  \\\n",
       "0      269,1  314,0  160,0  124,0  72,0  392,0  152,0  22,0  246,0  186,0   \n",
       "1      270,0  314,0  160,0  124,0  72,0  392,0  152,0  22,0  246,0  186,0   \n",
       "2      270,0  314,0  160,0  123,0  72,0  392,0  152,0  22,0  246,0  186,0   \n",
       "3      270,0  314,0  160,0  124,0  72,0  392,0  152,0  22,0  246,0  186,0   \n",
       "4      270,0  314,0  160,0  124,0  72,0  391,0  152,0  22,0  246,0  184,0   \n",
       "...      ...    ...    ...    ...   ...    ...    ...   ...    ...    ...   \n",
       "60880  270,0  314,0  158,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "60881  270,0  314,0  158,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "60882  270,0  314,0  158,0  126,0  72,0  386,0  154,0  22,0  246,0  186,0   \n",
       "60883  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "60884  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "\n",
       "      GWgam  MAcol  MAgam chri  \n",
       "0      16,0  168,0  250,0  0,0  \n",
       "1      16,0  168,0  250,0  0,0  \n",
       "2      16,0  168,0  250,0  0,0  \n",
       "3      16,0  168,0  250,0  0,0  \n",
       "4      16,0  168,0  250,0  0,0  \n",
       "...     ...    ...    ...  ...  \n",
       "60880  16,0  170,0  250,0  0,0  \n",
       "60881  16,0  170,0  250,0  0,0  \n",
       "60882  16,0  170,0  250,0  0,0  \n",
       "60883  16,0  170,0  250,0  0,0  \n",
       "60884  16,0  170,0  250,0  0,0  \n",
       "\n",
       "[60885 rows x 14 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2 = pd.read_csv('treemix/3L-free.new_treemix.txt.gz', sep=' ')\n",
    "df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "frames = [df_1, df_2]\n",
    "df = pd.concat(frames).reset_index().drop(columns=[\"index\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BFcol</th>\n",
       "      <th>BFgam</th>\n",
       "      <th>CIcol</th>\n",
       "      <th>GHcol</th>\n",
       "      <th>GHgam</th>\n",
       "      <th>GM-1</th>\n",
       "      <th>GM-2</th>\n",
       "      <th>GNcol</th>\n",
       "      <th>GNgam</th>\n",
       "      <th>GW</th>\n",
       "      <th>GWgam</th>\n",
       "      <th>MAcol</th>\n",
       "      <th>MAgam</th>\n",
       "      <th>chri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>270,0</td>\n",
       "      <td>312,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>393,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>268,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>392,2</td>\n",
       "      <td>150,4</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>181,5</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134870</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134871</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134872</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>158,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>386,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134873</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134874</th>\n",
       "      <td>270,0</td>\n",
       "      <td>314,0</td>\n",
       "      <td>160,0</td>\n",
       "      <td>126,0</td>\n",
       "      <td>72,0</td>\n",
       "      <td>394,0</td>\n",
       "      <td>154,0</td>\n",
       "      <td>22,0</td>\n",
       "      <td>246,0</td>\n",
       "      <td>186,0</td>\n",
       "      <td>16,0</td>\n",
       "      <td>170,0</td>\n",
       "      <td>250,0</td>\n",
       "      <td>0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>134875 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        BFcol  BFgam  CIcol  GHcol GHgam   GM-1   GM-2 GNcol  GNgam     GW  \\\n",
       "0       270,0  312,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "1       270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "2       268,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "3       268,0  314,0  160,0  126,0  72,0  393,0  154,0  22,0  246,0  186,0   \n",
       "4       268,0  314,0  160,0  126,0  72,0  392,2  150,4  22,0  246,0  181,5   \n",
       "...       ...    ...    ...    ...   ...    ...    ...   ...    ...    ...   \n",
       "134870  270,0  314,0  158,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "134871  270,0  314,0  158,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "134872  270,0  314,0  158,0  126,0  72,0  386,0  154,0  22,0  246,0  186,0   \n",
       "134873  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "134874  270,0  314,0  160,0  126,0  72,0  394,0  154,0  22,0  246,0  186,0   \n",
       "\n",
       "       GWgam  MAcol  MAgam chri  \n",
       "0       16,0  170,0  250,0  0,0  \n",
       "1       16,0  170,0  250,0  0,0  \n",
       "2       16,0  170,0  250,0  0,0  \n",
       "3       16,0  170,0  250,0  0,0  \n",
       "4       16,0  170,0  250,0  0,0  \n",
       "...      ...    ...    ...  ...  \n",
       "134870  16,0  170,0  250,0  0,0  \n",
       "134871  16,0  170,0  250,0  0,0  \n",
       "134872  16,0  170,0  250,0  0,0  \n",
       "134873  16,0  170,0  250,0  0,0  \n",
       "134874  16,0  170,0  250,0  0,0  \n",
       "\n",
       "[134875 rows x 14 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('treemix/chrom_3_chri_pca_pops.txt.gz', sep = \" \", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
